docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+  extends SinkNode(AdderNodeImp)(Seq(width))
+```
+
+Our adder node will take inputs from two `AdderDriverNode`s (the two addends)
+and pass one output (the sum) into the monitor. Because the number of inputs and
+outputs differ, we will make this node extend a `NexusNode`.
+
+`dFn` maps **incoming downward** parameters into **outgoing downward** parameters.
+Similarly, `uFn` maps **incoming upward** parameters into **outgoing upward** ones.
+
+```scala mdoc
+/** node for [[Adder]] (nexus) */
+class AdderNode(dFn: Seq[DownwardParam] => DownwardParam,
+                uFn: Seq[UpwardParam] => UpwardParam)(implicit valName: ValName)
+  extends NexusNode(AdderNodeImp)(dFn, uFn)
+```
+
+There are several other node types, which can be found in [Nodes.scala](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/diplomacy/Nodes.scala).
+
+### Creating the `LazyModule`s
+
+"Lazy" evaluation refers to deferring the evaluation of an expression until
+it is needed.
+
+Parameter negotiation in Diplomacy is done lazily, after the Diplomacy graph
+is created. Because of this, the concrete hardware that we want to parameterize
+must also be generated lazily. Diplomacy provides the `LazyModule` construct for
+writing lazily-evaluated hardware modules.
+
+Let's use `LazyModule` to define our `Adder`. Notice that the creation of the
+Diplomacy graph itself (in this case, the node), is non-lazy. The desired
+hardware for the module must be written inside `LazyModuleImp`.
+
+For this example, we expect one driver that should be driving two addends of an
+equal bit width into the adder, meaning we expect all of our downward flowing
+parameters to be equivalent. We also expect a sole monitor, so all of our
+upward flowing parameters should also be equivalent. Thus, we can require these
+properties in our `AdderNode` and simply pass on downward the first
+`DownwardParam` we see. Following the same logic, we can pass upward the
+first `UpwardParam` we see.
+
+```scala mdoc
+/** adder DUT (nexus) */
+class Adder(implicit p: Parameters) extends LazyModule {
+  val node = new AdderNode (
+    { case dps: Seq[DownwardParam] =>
+      require(dps.forall(dp => dp.width == dps.head.width))
+      dps.head
+    },
+    { case ups: Seq[UpwardParam] =>
+      require(ups.forall(up => up.width == ups.head.width))
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates addends each cycle. Note that each node
+passes on two identical `width` parameters to be consumed by the `Adder` and
+`AdderMonitor`.
+
+```scala mdoc
+/** driver (source) */
+class AdderDriver(width: Int)(implicit p: Parameters) extends LazyModule {
+  val nodeA = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
+  val nodeB = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
```

Comment:
It feels a bit weird to me to make one `val` per node in the graph inside this driver module, rather than having one node _per_ driver module and then making as many drivers as i want in the DUT, if you see what I mean? But it's not wrong per se, and I am not sure it is worth changing the figures and stuff over. It just means that this driver will only work for one specific graph, rather than being a tool that can be used as a component to make and test even bigger graphs. But maybe this change is something we could explore in future examples?

Reply:
I would also consider instead having one node for driving the monitors and one node for driving the adder, where each sequence length is based on the number of "drivers" you want to have in the DUT. That would let you write
```
val driver = AdderDriver(width = 8, num = 2)
adder.node :=* driver.adderNode
monitor.inputsNode :=* driver.monitorNode
monitor.sumNode := adder.node
```
But again this is trending towards a more complicated example

Reply:
I don't think it would be _too_ much effort to make these changes, and I do think it's better to have an example that aligns better with a more "canonical" diplomatic graph. I will give it a go!